Configuration header for list.h

Hello,

First of all I would like to say thank you for your library that is helpful a lot. I have a question/proposal - would you tell if it is worth adding a `LIST_CONFIG_H`, something like that:

```c
// ...
#ifdef LIST_CONFIG_H
#include LIST_CONFIG_H
#endif

#ifndef LIST_MALLOC
#define LIST_MALLOC malloc
#endif

#ifndef LIST_FREE
#define LIST_FREE free
#endif
// ...
```

That way we will be able to redefine required macros like `LIST_MALLOC` or `LIST_FREE` without dozen warnings:

```c
<command-line>: warning: implicit declaration of function 'sys_free' [-Wimplicit-function-declaration]
list/src/list.c:40:5: note: in expansion of macro 'LIST_FREE'
   40 |     LIST_FREE(curr);
      |     ^~~~~~~~~
list/src/list_iterator.c: In function 'list_iterator_new_from_node':
<command-line>: warning: implicit declaration of function 'sys_malloc'; did you mean 'malloc'? [-Wimplicit-function-declaration]
list/src/list_iterator.c:31:16: note: in expansion of macro 'LIST_MALLOC'
   31 |   if (!(self = LIST_MALLOC(sizeof(list_iterator_t))))
      |                ^~~~~~~~~~~
list/src/list_iterator.c:31:14: warning: assignment to 'list_iterator_t *' {aka 'struct <anonymous> *'} from 'int' makes pointer from integer without a cast [-Wint-conversion]
   31 |   if (!(self = LIST_MALLOC(sizeof(list_iterator_t))))
      |              ^
list/src/list_iterator.c: In function 'list_iterator_destroy':
<command-line>: warning: implicit declaration of function 'sys_free' [-Wimplicit-function-declaration]
list/src/list_iterator.c:60:3: note: in expansion of macro 'LIST_FREE'
   60 |   LIST_FREE(self);
      |   ^~~~~~~~~
```

If that seems reasonable to you, I'll be able to elaborate on a PR for that change.

Thank you.

--- a/src/list.c
+++ b/src/list.c
@@ -12,7 +12,7 @@
  */
 
 list_t *
-list_new() {
+list_new(void) {
   list_t *self;
   if (!(self = LIST_MALLOC(sizeof(list_t))))
     return NULL;

--- a/src/list.h
+++ b/src/list.h
@@ -83,7 +83,7 @@ list_node_new(void *val);
 // list_t prototypes.
 
 list_t *
-list_new();
+list_new(void);
 
 list_node_t *
 list_rpush(list_t *self, list_node_t *node);

--- a/src/list.h
+++ b/src/list.h
@@ -19,6 +19,13 @@ extern "C" {
 #define LIST_VERSION "0.0.5"
 
 // Memory management macros
+#ifdef LIST_CONFIG_H
+#define _STR(x) #x
+#define STR(x) _STR(x)
+#include STR(LIST_CONFIG_H)
+#undef _STR
+#undef STR
+#endif
 
 #ifndef LIST_MALLOC
 #define LIST_MALLOC malloc

--- a/src/list.h
+++ b/src/list.h
@@ -5,8 +5,8 @@
 // Copyright (c) 2010 TJ Holowaychuk <tj@vision-media.ca>
 //
 
-#ifndef LIST_H
-#define LIST_H
+#ifndef __CLIBS_LIST_H__
+#define __CLIBS_LIST_H__
 
 #ifdef __cplusplus
 extern "C" {
@@ -127,4 +127,4 @@ list_iterator_destroy(list_iterator_t *self);
 }
 #endif
 
-#endif /* LIST_H */
+#endif /* __CLIBS_LIST_H__ */

